### PR TITLE
Replace Unpaginated endpoints with unified metadata for sync

### PR DESCRIPTION
**Pull request by rochacbruno**
- Remove Unpaginated /all/ endpoints
- Adds a Paginator with bigger limit
- Adds /collection/metadata/ endpoint with unified metadata for
  collection and its versions
- Changed sync to use this `optimized` downloader using 1000 limit per
  pagination

output example: https://gist.github.com/rochacbruno/df9c1c3e2653d1e9592cd8f650de2239

this is a temporary solution to be replaced by #553 which will pack the collections/metadata/ in a publication.
related to: https://github.com/ansible/galaxy_ng/pull/726

fixes: #8439

### CODE REVIEW

**Review by fao89 — COMMENTED** [review]
This is looking good!

File: pulp_ansible/app/tasks/collections.py [hunk 1]
```diff
@@ -530,16 +527,12 @@ async def _fetch_paginated_collection_metadata(self, name, namespace, requiremen
...
         loop = asyncio.get_event_loop()
 
-        if self._unpaginated_collection_metadata[namespace][name]["deprecated"]:
-            self.deprecations |= Q(namespace=namespace, name=name)
-
-        all_versions_of_collection = self._unpaginated_collection_version_metadata[namespace][name]
-
-        for col_version_metadata in all_versions_of_collection:
+        for col_version_metadata in versions:
+            __import__('sdb').set_trace()
```
Comment (fao89) [inline pulp_ansible/app/tasks/collections.py:535]: I removed this

File: pulp_ansible/app/tasks/collections.py [hunk 2]
```diff
@@ -663,17 +600,33 @@ async def _find_all_collections(self):
...
                     name=".".join([namespace, name]),
                     version="*",
                     source=None,
                 )
-                tasks.append(loop.create_task(self._fetch_collection_metadata(requirements_file)))
+
+                if collection['deprecated']:
+                    self.deprecations |= Q(namespace=namespace, name=name)
+
+                if 'versions' not in collection:
```
Comment (fao89) [inline pulp_ansible/app/tasks/collections.py:619]:
```suggestion
                self._api_version = api_version
                if 'versions' not in collection:
```

**Review by fao89 — COMMENTED** [review]
with these 2 minor changes I was able to have the tests passing:
```
(pulp) [vagrant@pulp3-source-fedora33 pulp_ansible]$ pytest pulp_ansible/tests/functional/                               
===================================== test session starts =====================================
platform linux -- Python 3.9.2, pytest-6.2.3, py-1.10.0, pluggy-0.13.1
rootdir: /home/vagrant/devel/pulp_ansible
collected 74 items                                                                            

pulp_ansible/tests/functional/api/test_export.py .                                      [  1%]
pulp_ansible/tests/functional/api/test_import.py ..                                     [  4%]
pulp_ansible/tests/functional/api/collection/test_crud_collection_versions.py .         [  5%]
pulp_ansible/tests/functional/api/collection/test_remote.py .....                       [ 12%]
pulp_ansible/tests/functional/api/collection/test_sync.py ...........                   [ 27%]
pulp_ansible/tests/functional/api/collection/v2/test_sync.py s..........                [ 41%]
pulp_ansible/tests/functional/api/collection/v2/test_upload.py .                        [ 43%]
pulp_ansible/tests/functional/api/collection/v3/test_collection.py ....s.               [ 51%]
pulp_ansible/tests/functional/api/collection/v3/test_deprecation.py .                   [ 52%]
pulp_ansible/tests/functional/api/collection/v3/test_serializers.py ..                  [ 55%]
pulp_ansible/tests/functional/api/collection/v3/test_sync.py ....ssss                   [ 66%]
pulp_ansible/tests/functional/api/role/test_crd_distribution.py .........               [ 78%]
pulp_ansible/tests/functional/api/role/test_crud_remotes.py ........                    [ 89%]
pulp_ansible/tests/functional/api/role/test_download_content.py s                       [ 90%]
pulp_ansible/tests/functional/api/role/test_publish.py s                                [ 91%]
pulp_ansible/tests/functional/api/role/test_sync.py ..                                                                                                                                 [ 94%]
pulp_ansible/tests/functional/cli/test_collection_install.py .                                                                                                                         [ 95%]
pulp_ansible/tests/functional/cli/test_collection_upload.py ..                                                                                                                         [ 98%]
pulp_ansible/tests/functional/cli/test_role_install.py .                                                                                                                               [100%]

====================================================================================== warnings summary ======================================================================================
pulp_ansible/tests/functional/api/collection/v2/test_upload.py::UploadCollectionTestCase::test_collection_upload
pulp_ansible/tests/functional/api/collection/v3/test_collection.py::test_collection_upload_repeat
  /usr/local/lib/pulp/lib64/python3.9/site-packages/urllib3/connectionpool.py:1013: InsecureRequestWarning: Unverified HTTPS request is being made to host 'galaxy.ansible.com'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
    warnings.warn(

pulp_ansible/tests/functional/api/collection/v2/test_upload.py::UploadCollectionTestCase::test_collection_upload
pulp_ansible/tests/functional/api/collection/v3/test_collection.py::test_collection_upload_repeat
  /usr/local/lib/pulp/lib64/python3.9/site-packages/urllib3/connectionpool.py:1013: InsecureRequestWarning: Unverified HTTPS request is being made to host 'ansible-galaxy.s3.amazonaws.com'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=================================================================== 66 passed, 8 skipped, 4 warnings in 723.06s (0:12:03) ====================================================================

```
Note: I only have pulp_ansible on my environment, I'm wondering if having pulp_ansible + galaxy_ng may cover some errors

**Comment by fao89** [comment]
I've tested your PR with https://github.com/pulp/pulp_ansible/pull/559
and I was able to sync from this
image

**Comment by fao89** [comment]
Lint is breaking due to black formatting and because it needs a changelog entry on CHANGES dir

File: pulp_ansible/app/tasks/collections.py (hunk 2 above)
Comment (rochacbruno) [reply pulp_ansible/app/tasks/collections.py]: I added this outside the loop.

**Review by fao89 — APPROVED** [review]
LGTM!

**Comment by fao89** [comment]
@gerrod3 since you worked recently with collection tasks, please take a look!

**Comment by rochacbruno** [comment]
@gerrod3 lets close it and go with #562 ?

**Pull request closed without merging**